File: VB/WinSolution.Win/WinApplication.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Win
Imports DevExpress.Data.Filtering

Namespace WinSolution.Win
	Partial Public Class WinSolutionWindowsFormsApplication
		Inherits WinApplication
		Public Sub New()
			InitializeComponent()
			AddHandler CreateCustomCollectionSource, AddressOf CreateCollectionSourceEvent
		End Sub

		Private Sub WinSolutionWindowsFormsApplication_DatabaseVersionMismatch(ByVal sender As Object, ByVal e As DatabaseVersionMismatchEventArgs) Handles MyBase.DatabaseVersionMismatch
			e.Updater.Update()
			e.Handled = True
		End Sub

		Public Sub CreateCollectionSourceEvent(ByVal sender As Object, ByVal args As CreateCustomCollectionSourceEventArgs)
			If args.ListViewID.EndsWith("_Deleted") Then
			   args.CollectionSource = New DeletedObjectsCollectionSource(args.ObjectSpace, args.ObjectType)
			End If
		End Sub
	End Class
	Public Class DeletedObjectsCollectionSource
		Inherits CollectionSource
		Public Shared ReadOnly DeletedObjectsCriteria As CriteriaOperator = CriteriaOperator.Parse("GCRecord is not null")
		Public Sub New(ByVal objectSpace As IObjectSpace, ByVal objectType As Type)
			MyBase.New(objectSpace, objectType)
		End Sub
		Protected Overrides Sub ApplyCriteriaCore(ByVal criteria As CriteriaOperator)
			MyBase.ApplyCriteriaCore(criteria And DeletedObjectsCriteria)
		End Sub
		Protected Overrides Function CreateCollection() As Object
			Dim collection As XPCollection = TryCast(MyBase.CreateCollection(), XPCollection)
			If collection IsNot Nothing Then
				collection.SelectDeleted = True
			End If
			Return collection
		End Function
	End Class
End Namespace

File: CS/WinSolution.Win/WinApplication.cs
using System;
using DevExpress.Xpo;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Win;
using DevExpress.Data.Filtering;

namespace WinSolution.Win {
    public partial class WinSolutionWindowsFormsApplication : WinApplication {
        public WinSolutionWindowsFormsApplication() {
            InitializeComponent();
            CreateCustomCollectionSource += new EventHandler<CreateCustomCollectionSourceEventArgs>(CreateCollectionSourceEvent);
        }

        private void WinSolutionWindowsFormsApplication_DatabaseVersionMismatch(object sender, DatabaseVersionMismatchEventArgs e) {
            e.Updater.Update();
            e.Handled = true;
        }

        public void CreateCollectionSourceEvent(object sender, CreateCustomCollectionSourceEventArgs args) {
            if(args.ListViewID.EndsWith("_Deleted")) {
               args.CollectionSource = new DeletedObjectsCollectionSource(args.ObjectSpace, args.ObjectType);
            }
        }
    }
    public class DeletedObjectsCollectionSource : CollectionSource {
        public static readonly CriteriaOperator DeletedObjectsCriteria = CriteriaOperator.Parse("GCRecord is not null");
        public DeletedObjectsCollectionSource(ObjectSpace objectSpace, Type objectType) : base(objectSpace, objectType) { }
        protected override void ApplyCriteriaCore(CriteriaOperator criteria) {
            base.ApplyCriteriaCore(criteria & DeletedObjectsCriteria);
        }
        protected override object CreateCollection() {
            XPCollection collection = base.CreateCollection() as XPCollection;
            if (collection != null) {
                collection.SelectDeleted = true;
            }
            return collection;
        }
    }
}
